samila/functions.py

```diff
@@ -159,6 +162,33 @@ def filter_color(color, bgcolor):
     return color, bgcolor
 
 
+def filter_cmap(cmap):
+    """
+    Filter given cmap.
+
+    :param cmap: color map
+    :type cmap: matplotlib.colors.Colormap or list of colors
+    :return: filtered version of cmap
+    """
+    if isinstance(cmap, str):
+        cmap = cm.get_cmap(cmap, 256)
+    if isinstance(cmap, matplotlib.colors.ListedColormap):
+        return cmap
+    if isinstance(cmap, (matplotlib.colors.LinearSegmentedColormap)):
+        cmap = cmap(range(DEFAULT_CMAP_RANGE))
+    if isinstance(cmap, matplotlib.colors.Colormap):
+        cmap = cm.get_cmap(cmap.__getattribute__("name"))
+    if isinstance(cmap, matplotlib.colors.ListedColormap):
+        return cmap
+    if isinstance(cmap, (matplotlib.colors.LinearSegmentedColormap)):
+        cmap = cmap(range(DEFAULT_CMAP_RANGE))
+        return ListedColormap(cmap)
+    if isinstance(cmap, list):
+        cmap = list(map(select_color, cmap))
+        return ListedColormap(cmap)
```

Comment:
The repeated part is because below check
```python
    if isinstance(cmap, matplotlib.colors.Colormap):
        cmap = cm.get_cmap(cmap.__getattribute__("name"))
```
It changes the camp and we require doing the above check again, on the other hand, it'd raise some issues putting them after this line because the `matplotlib.colors.ListedColormap` and `matplotlib.colors.LinearSegmentedColormap` both extend the `matplotlib.colors.Colormap`.

Reply:
This function is too complex!
We have two options:

1. Using `type` instead of `isinstance` to identify the exact type (not superclass/subclass)
For example:
```pycon
>>> import matplotlib
>>> from matplotlib import cm
>>> g = cm.get_cmap("viridis")
>>> isinstance(g, matplotlib.colors.Colormap)
True
>>> isinstance(g, matplotlib.colors.ListedColormap)
True
>>> type(g) == matplotlib.colors.Colormap
False
>>> type(g) == matplotlib.colors.ListedColormap
True
```

2. Using `hasattr` function.
For example:
```python3
if hasattr(cmap, "name"): # This will cover all colormap types have `name` attribute.
    cmap = cm.get_cmap(cmap.name)     
if isinstance(cmap, str):
    cmap = cm.get_cmap(cmap, 256)
if isinstance(cmap, matplotlib.colors.ListedColormap):
    return cmap
if isinstance(cmap, (matplotlib.colors.LinearSegmentedColormap)):
    cmap = cmap(range(DEFAULT_CMAP_RANGE))
    return ListedColormap(cmap)
if isinstance(cmap, list):
    cmap = list(map(select_color, cmap))
    return ListedColormap(cmap)
 ```

Reply:
Thank you. I think I would choose the second one.

Reply:
Unfortunately it seems that:
```python
>>> cmap = [[0.7, 0.2, 0.2, 1], [0.6, 0.2, 0.2, 1], [0.3, 0.2, 0.2, 1], [0.2, 0.2, 0.2, 1]]
>>> from matplotlib.colors import ListedColormap
>>> from samila import *
>>> g = GenerativeImage()
>>> g.generate()
>>> x = ListedColormap(cmap)
>>> hasattr(x, "name")
True
>>> x.name
'from_list'
```